### PR TITLE
EDMM integration into Eclipse Winery

### PR DESCRIPTION
**Pull request by wikilele**
The pull request integrates EDMM into winery.
The module `org.eclipse.winery.edmm` has been added and it wraps the edmm core.
Some API endpoints have been implemented in the `org.eclipse.winery.repository.rest` module allowing to access the edmm module utilities via REST interface. 
The edmm UI in the topologymodeler is the same as before, but now it allows to call also the `transform` function of the edmm core.

### CODE REVIEW

**Comment by genie-winery** [comment]
Can one of the admins verify this patch?

**Comment by wikilele** [comment]
@miwurster is going to review this pull request.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/EdmmTransformationFramework.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.eclipse.winery.edmm.plugins.PluginManager;
+import org.eclipse.winery.edmm.utils.ZipUtility;
+
+import io.github.edmm.core.parser.EntityGraph;
+import io.github.edmm.model.DeploymentModel;
+import io.github.edmm.core.plugin.PluginService;
+import io.github.edmm.core.transformation.TransformationService;
+
+public class EdmmTransformationFramework {
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/EdmmTransformationFramework.java:18]: Rename to `TransformationManager`.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+import lombok.Getter;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.w3c.dom.Node;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PluginManager{
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java:36]: Apply the appropriate Winery source code formatting. See dev guide.

Comment (miwurster) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java:36]: On top of that, pls make sure that you add your name to the NOTICE and .mailmap file. Further, you have to sign the ECA by Eclipse: https://www.eclipse.org/legal/ECA.php

File: org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/EdmmDecisionSupportSystemTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.List;
+
+import io.github.edmm.core.parser.EntityGraph;
+import io.github.edmm.model.PluginSupportResult;
+
+import org.eclipse.winery.edmm.plugins.PluginManager;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class  EdmmDecisionSupportSystemTest extends TestEdmm{
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/EdmmDecisionSupportSystemTest.java:26]: Refactor to a general-purpose `EdmmTests` class.

File: org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/EdmmTransformationFrameworkTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import io.github.edmm.core.parser.EntityGraph;
+
+import org.apache.commons.io.FileUtils;
+import org.eclipse.jgit.api.Git;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class EdmmTransformationFrameworkTest extends TestEdmm{
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/EdmmTransformationFrameworkTest.java:18]: Move to `EdmmTests` class.

File: org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/TestEdmm.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.io.File;
+
+import io.github.edmm.core.parser.EntityGraph;
+import io.github.edmm.model.DeploymentModel;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class TestEdmm {
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/TestEdmm.java:11]: Move to `EdmmTests` class.

File: org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/plugins/PluginManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.DocumentBuilder;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+
+public class  PluginManagerTest{
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/test/java/org/eclipse/winery/edmm/plugins/PluginManagerTest.java:27]: Move to `EdmmTests` class.

File: org.eclipse.winery.edmm/src/test/resources/petclinic_IaaS.yml [hunk 7]
```diff
@@ -0,0 +1,125 @@
+---
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/test/resources/petclinic_IaaS.yml:1]: Rename file to `petclinic-iaas.yml`.

File: org.eclipse.winery.edmm/src/test/resources/petclinic_deployment.yml [hunk 8]
```diff
@@ -0,0 +1,207 @@
+---
```
Comment (miwurster) [inline org.eclipse.winery.edmm/src/test/resources/petclinic_deployment.yml:1]: Rename to `petclinic-paas.yml`.

File: org.eclipse.winery.repository.rest/pom.xml [hunk 9]
```diff
@@ -50,13 +52,13 @@
         <dependency>
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-databind</artifactId>
-            <version>2.8.6</version>
+            <version>${com.fasterxml.jackson.core}</version>
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/pom.xml:55]: We are not allowed to change this. Revert.

File: org.eclipse.winery.repository.rest/pom.xml [hunk 10]
```diff
@@ -50,13 +52,13 @@
...
             <scope>compile</scope>
         </dependency>
         <dependency>
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-core</artifactId>
-            <version>2.8.6</version>
+            <version>${com.fasterxml.jackson.core}</version>
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/pom.xml:61]: We are not allowed to change this. Revert.

File: org.eclipse.winery.repository.rest/pom.xml [hunk 11]
```diff
@@ -209,6 +211,24 @@
...
+        <dependency>
+            <groupId>org.jgrapht</groupId>
+            <artifactId>jgrapht-core</artifactId>
+            <version>1.1.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-yaml</artifactId>
+            <!-- TODO maybe it's better to have com.fasterxml.jackson.dataformat -->
+            <version>${com.fasterxml.jackson.core}</version>
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/pom.xml:229]: Use same version as in Winery.

File: org.eclipse.winery.repository.rest/pom.xml [hunk 12]
```diff
@@ -209,6 +211,24 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.jgrapht</groupId>
+            <artifactId>jgrapht-core</artifactId>
+            <version>1.1.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-yaml</artifactId>
+            <!-- TODO maybe it's better to have com.fasterxml.jackson.dataformat -->
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/pom.xml:228]: Why better?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/RestUtils.java [hunk 13]
```diff
@@ -324,7 +332,7 @@ public static Response getEdmmModel(TServiceTemplate element) {
             .entity(stringWriter.toString())
             .build();
     }
-
+
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/RestUtils.java:335]: You definitely have to load the Winery source code formatting rules!

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java [hunk 14]
```diff
@@ -35,6 +35,7 @@
 import javax.ws.rs.core.UriInfo;
 import javax.xml.bind.JAXBException;
 import javax.xml.bind.Unmarshaller;
+//import javax.xml.crypto.dsig.TransformService; unused import
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java:38]: What? Remove it if it's unused! Also check our `checkstyle` rules!

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java [hunk 15]
```diff
@@ -506,7 +508,12 @@ public Response updateDefinitions(InputStream requestBodyStream) {
             return Response.ok().entity(validationError).build();
         }
     }
-
+    
+    @Path("edmm")
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java:512]: Why is this in the abstract class? Shouldn't this be added to the `ServiceTemplateInstanceResource`?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmAPI.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
+package org.eclipse.winery.repository.rest.edmm;
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmAPI.java:1]: Strange naming. Use the Winery convention -> `EdmmResource`.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmAPI.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+        DeploymentModel deploymentModel = DeploymentModel.of(yml);
+
+        return deploymentModel.getGraph(); 
+    }
+        
+    @GET
+    @Path("/")
+    @Produces(MediaType.TEXT_PLAIN)
+    public Response hello() {
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmAPI.java:65]: Really? :-)

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmAPITest.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
+/*******************************************************************************
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmAPITest.java:1]: Rename to `EdmmResourceTest`

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmAPITest.java [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+    
+    /**
+     * slashes must be encoded with %252F
+     */
+    private static final String PETCLINIC_TOPOLOGY_CLOUD = "servicetemplates/https%253A%252F%252Fedmm.uni-stuttgart.de%252Fservicetemplates/PetClinic-Cloud/";
+    private static final String PETCLINIC_TOPOLOGY_IAAS = "servicetemplates/https%253A%252F%252Fedmm.uni-stuttgart.de%252Fservicetemplates/PetClinic-IaaS/";
+    
+    
+    @Test
+    public void testHello() throws Exception {
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmAPITest.java:41]: Oh my :-)

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmAPITestSetUp.java [hunk 20]
```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmAPITestSetUp.java:1]: Move this to the `EdmmResourceTest` class!

**Comment by miwurster** [comment]
Also, resolve the conflicts with `master`.

File: org.eclipse.winery.edmm/pom.xml [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+            <id>jitpack.io</id>
+            <url>https://jitpack.io</url>
+        </repository>
+    </repositories>
+    
+    <dependencies>
+        <dependency>
+            <groupId>com.github.ust-edmm.transformation-framework</groupId>
+            <artifactId>edmm-core</artifactId>
+            <version>7cf02a73e7ad0e8e09ba2668839123b71e538577</version>
```
[the remark this answers is not in the document]
Comment (wikilele) [reply org.eclipse.winery.edmm/pom.xml:42]: I tried to put `master` but jitpack seems not rebuilding the edmm-core. The new master it's not downloaded; maven uses an old version causing an error because it doesn't found a class.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java (hunk 2 above)
Comment (wikilele) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java:36]: Can I just add myself to those files simply writing my name and email?

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java (hunk 15 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java:512]: Do you mean `ServiceTemplateResource.java` ?

File: org.eclipse.winery.edmm/pom.xml (hunk 21 above)
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.edmm/pom.xml:42]: Ok, then we cannot merge until this has been resolved. We have to wait for a release of the transformation framework then.

File: org.eclipse.winery.repository.rest/pom.xml [hunk 22]
```diff
@@ -209,6 +209,23 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.jgrapht</groupId>
+            <artifactId>jgrapht-core</artifactId>
+            <version>1.1.0</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-yaml</artifactId>
+            <version>2.8.6</version>
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/pom.xml:226]: make sure you use the same version as Winery currently uses.

File: org.eclipse.winery.repository.rest/pom.xml [hunk 23]
```diff
@@ -209,6 +209,23 @@
...
+        <!-- begin of EDMM dependencies -->
+        <dependency>
+            <groupId>org.eclipse.winery</groupId>
+            <artifactId>org.eclipse.winery.edmm</artifactId>
+            <version>2.0.0-SNAPSHOT</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jgrapht</groupId>
+            <artifactId>jgrapht-core</artifactId>
+            <version>1.1.0</version>
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/pom.xml:221]: make sure you use the same version as Winery currently uses.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * Winery at the moment doesn't create a valid EntityGraph, but a graph that will become valid only after 
+     * the conversion to a YAML file. It works like this because originally only the YAML string was needed.
+     * 
+     * Now we use a function that takes the 'invalid' EntityGraph converts it to a valid YAML and then gets back
+     * a valid EntityGraph.
+     * 
+     */
+    private EntityGraph correctGraph(EntityGraph graph) {
```
Comment (miwurster) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java:51]: Shouldn't be part of this PR.

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java (hunk 15 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java:511]: If everything related to a service template instance is handled in `ServiceTemplateResource.java`, yes :-)

File: org.eclipse.winery.repository.rest/pom.xml (hunk 22 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml:226]: In the root pom.xml the version specified for `com.fasterxml.jackson.core.jackson-annotations` is `2.8.0`. In this pom.xml the version used for `jackson-databind` and `jackson-core` is `2.8.6` .

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java (hunk 24 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java:51]: But without that correction it doesn't work. Where should I put that? In the edmm-core?

File: org.eclipse.winery.edmm/pom.xml (hunk 21 above)
[the remark this answers is not in the document]
Comment (wikilele) [reply org.eclipse.winery.edmm/pom.xml:42]: Can't you access jitpack web site and delete the master build while keeping the last release?

File: org.eclipse.winery.repository.rest/pom.xml (hunk 23 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml:221]: I have tried using the version specified in the root pom.xml file (1.3.1) but it fail giving as error:
`java.lang.NoClassDefFoundError: org/jgrapht/EdgeFactory`

File: org.eclipse.winery.repository.rest/pom.xml (hunk 22 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/pom.xml:226]: Versions need to be aligned.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 23 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/pom.xml:221]: Solve this, please. Check where the NoClassDefFoundError occurs and migrate it. I suggest du try using the latest possible version (probably the one that comes with the transformation framework)

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java (hunk 24 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java:51]: Ah, misunderstood the method. It's fine

File: org.eclipse.winery.edmm/pom.xml (hunk 21 above)
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.edmm/pom.xml:42]: We are anyhow not allowed to use the master build. We have to wait for the next release of the transformation framework.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 23 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml:221]: The transformation-framework used the version `1.1.0`, the one I putted here.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 22 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml:226]: To which one? The version numbers I listed above are there not because of me. I can't touch them if I understood well. I putted here the version 2.8.6 because it's the same used in `jackson-databind` and `jackson-core`.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 23 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/pom.xml:221]: Then upgrade the transformation framework to have it working with 1.3.1. It's a MUST HAVE to align these versions.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 22 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/pom.xml:226]: Try to use version that is already "in use" by Winery.

Comment (miwurster) [reply org.eclipse.winery.repository.rest/pom.xml:226]: So, `2.8.6` everywhere, also in the `<com.fasterxml.jackson.core.jackson-annotations>` property.

Comment (miwurster) [reply org.eclipse.winery.repository.rest/pom.xml:226]: Even better, change the Maven property to `<com.fasterxml.jackson>`, set it to `2.8.6` and adapt any pom.xml file to reflect this property with `${com.fasterxml.jackson}` in the respective `<version>` tags.

Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml:226]: that was what I did before reverting to that. Ok, no problem I'll reset the pom.xml property.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 23 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml]: this should be solved. I updated the version of jgrapht on the transformation-framework and setted here the version used by winery.

File: org.eclipse.winery.repository.rest/pom.xml (hunk 22 above)
Comment (wikilele) [reply org.eclipse.winery.repository.rest/pom.xml]: Now there is a property `<com.fasterxml.jackson>` in the root pom.xml. I have changed all the dependencies to jackson so that now they  use this property.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+public class ZipUtility {
```
Comment (Vogel612) [inline org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:13]: There's overlap here to the Zip methods exposed by `org.eclipse.winery.yaml.common.Utils` that allow zipping and unzipping as well as the Zipping and unzipping in FilebasedRepository and quite probably other places in the codebase...

It's probably worth checking for a way to unify these, because zipping can be non-obvious.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * The function is used to zip a directory in a file
+     * @param sourcePath the path of the directory
+     * @param destPath the destination path of the zip file,
+     *                 if a zip file already exists at the destination path, it will be overwritten
+     *
+     * https://stackoverflow.com/questions/15968883/how-to-zip-a-folder-itself-using-java
+     */
+    public static ZipOutputStream pack(Path sourcePath, Path destPath) throws IOException {
```
Comment (Vogel612) [inline org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:23]: a non-obvious edge-case here is when `destPath.beginsWith(sourcePath)`, as in, the `destPath` is part of the tree that gets walked with `Files.walk`. In that case the code will probably deadlock without an error messge.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+            
+            // creating the directories in which the newFile shoud be stored
+            newFile.toPath().getParent().toFile().mkdirs();
+                        
+            FileOutputStream fos = new FileOutputStream(newFile);
+            int len;
+            while ((len = zipInputStream.read(buffer)) > 0) {
+                fos.write(buffer, 0, len);
+            }
+            fos.close();
```
Comment (Vogel612) [inline org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:80]: This should be replaceable by using 

```suggestion
            Path newFile = destDirPath.resolve(zipEntry.getName());
            Files.createDirectories(newFile.getParent());
            Files.copy(zipInputStream, newFile);
```

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java (hunk 25 above)
Comment (wikilele) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:13]: Yes, you are absolutely right. I can't work on this right now, but I'll check in the next days if this class can be removed or unified with the other zipping functions around the code. 

The idea is to just use the `org.eclipse.winery.yaml.common.Utils` zip function and remove the other copies, is this right?

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java (hunk 26 above)
Comment (wikilele) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:23]: If we want a general zip function we need to consider this case too.
 At the moment I think it won't happen a situation like this. When the function is called the `destPath` is never a sub tree of the `sourcePath`.

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java (hunk 27 above)
Comment (wikilele) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:80]: Yes, your code is better. It can be committed I think..

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java (hunk 25 above)
Comment (Vogel612) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:13]: I don't have an exhaustive overview where in winery Zip handing is done, but I assume there are multiple different usage patterns that are not that easy to unify. The idea is to centralize the Zip handling into a single utility class. That class probably *should* be in `org.eclipse.winery.common`.

That change could be a bit out-of-scope for this PR, though... I'm not sure on the constraints that are at play here.

Comment (wikilele) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/utils/ZipUtility.java:13]: Ok, I will check if I can get rid of this "duplicated" code using one zip utility already implemented.
If it's not possible I'll apply the changes suggested below, but I won't unify the zip functions since it's not the scope of this PR. 
It's anyway better not to mark this conversation as resolved because this zip utility problem needs to be fixed in the right way.

File: org.eclipse.winery.edmm/pom.xml [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+            <id>jitpack.io</id>
+            <url>https://jitpack.io</url>
+        </repository>
+    </repositories>
+    
+    <dependencies>
+        <dependency>
+            <groupId>com.github.ust-edmm.transformation-framework</groupId>
+            <artifactId>edmm-core</artifactId>
+            <version>3dbc40f1c0e3bfc14e8ad006ef7cf0ffb8b01872</version>
```
Comment (wikilele) [inline org.eclipse.winery.edmm/pom.xml:42]: The commit hash here should be removed and replaced with the release version.

**Review by miwurster — APPROVED** [review]

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java [hunk 29]
```diff
@@ -1,5 +1,5 @@
 /********************************************************************************
- * Copyright (c) 2012-2020 Contributors to the Eclipse Foundation
+ * Copyright (c) 2012-2021 Contributors to the Eclipse Foundation
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java:2]: There are no real changes in this file, revert them.

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmResourceTest.java [hunk 30]
```diff
@@ -0,0 +1,224 @@
...
+    protected static final String CHECK_MODEL = "edmm/dss/check-model-support";
+
+    /**
+     * slashes must be encoded with %252F
+     */
+    private static final String PETCLINIC_TOPOLOGY_CLOUD = "servicetemplates/https%253A%252F%252Fedmm.uni-stuttgart.de%252Fservicetemplates/PetClinic-Cloud/";
+    private static final String PETCLINIC_TOPOLOGY_IAAS = "servicetemplates/https%253A%252F%252Fedmm.uni-stuttgart.de%252Fservicetemplates/PetClinic-IaaS/";
+
+    private JSONArray callCheckModel(String topology) throws JSONException {
+        String responseJson = this.start()
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmResourceTest.java:52]: Isn't this already covered by `assertGet()`?

File: org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmResourceTest.java [hunk 31]
```diff
@@ -0,0 +1,224 @@
...
+            .filter(Files::isRegularFile)
+            .collect(Collectors.toList());
+
+        Assert.assertEquals(12, deploymentFiles.size());
+    }
+
+    @Test
+    public void testGetOneToOneMap() throws Exception {
+        this.setRevisionTo("c606d96aeee4dc086e018fb30fb0198a74384f06");
+        String oneToOneMapAsString = this.start()
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/test/java/org/eclipse/winery/repository/rest/edmm/EdmmResourceTest.java:171]: see above

**Review by lharzenetter — CHANGES_REQUESTED** [review]

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java [hunk 32]
```diff
@@ -0,0 +1,264 @@
...
+            .type(MediaType.APPLICATION_JSON)
+            .entity(reverseOneToOneMap)
+            .build();
+    }
+
+    @POST
+    @Consumes(MediaType.TEXT_PLAIN)
+    @Path("create-placeholders-scripts")
+    public Response createPlaceholders(String componentType, @Context UriInfo uriInfo) throws IOException {
+        final String uniStuttgart = "https://edmm.uni-stuttgart.de/";
```
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java:169]: Those are just defaults

File: org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java [hunk 33]
```diff
@@ -0,0 +1,116 @@
...
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+
+import io.github.edmm.core.plugin.PluginService;
+import io.github.edmm.core.plugin.TransformationPlugin;
+import lombok.Getter;
```
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.edmm/src/main/java/org/eclipse/winery/edmm/plugins/PluginManager.java]: Fixed

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/edmmTransformationCheck/edmm-replacement-rules/edmm-replacement-rules.component.html [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+        <span> {{pluginName}}'s rules </span> <br>
+        <div [ngClass]="getColorClass()">
+            <span> {{components}} </span>
+        </div>
+        replace it with the topology:
+        <div class="applicable">
+            <pre>{{toTopology}}</pre>
+        </div>
+    <button class="dark-button-style" (click)="prevRule()"> prev </button>
+    <button class="dark-button-style" (mousedown)="applyRule()"> apply </button>
```
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.frontends/app/topologymodeler/src/app/edmmTransformationCheck/edmm-replacement-rules/edmm-replacement-rules.component.html]: Fixed

File: .idea/runConfigurations/Winery_Server.xml [hunk 35]
```diff
@@ -11,7 +11,7 @@
     </deployment>
     <server-settings>
       <option name="BASE_DIRECTORY_NAME" value="Unnamed_winery_3" />
-      <option name="JNDI_PORT" value="26416" />
+      <option name="JNDI_PORT" value="8686" />
```
[the remark this answers is not in the document]
Comment (miwurster) [reply .idea/runConfigurations/Winery_Server.xml]: Fixed

File: org.eclipse.winery.edmm/pom.xml [hunk 36]
```diff
@@ -52,6 +52,12 @@
             <artifactId>org.eclipse.winery.repository</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
```
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.edmm/pom.xml]: Fixed

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java [hunk 37]
```diff
@@ -0,0 +1,264 @@
...
+public class EdmmResource {
+
+    private final TServiceTemplate element;
+
+    public EdmmResource(TServiceTemplate element) {
+        this.element = element;
+    }
+
+    @GET
+    @Path("dss/check-model-support")
```
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java]: Fixed

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java (hunk 29 above)
Comment (miwurster) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/_support/AbstractComponentInstanceResource.java]: Fixed

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java [hunk 38]
```diff
@@ -0,0 +1,264 @@
...
+            artifactApiData.interfaceName = lifecycleName;
+            artifactApiData.operationName = operation;
+
+            boolean implementationResourceExists = nodeTypeImplementationResource.getImplementationArtifacts()
+                .getAllArtifactResources()
+                .stream()
+                .anyMatch(artifactResource -> {
+                    TImplementationArtifacts.ImplementationArtifact implementationArtifact = artifactResource.getImplementationArtifact();
+                    return operation.equals(implementationArtifact.getName()) &&
+                        operation.equals(implementationArtifact.getOperationName()) &&
```
[the remark this answers is not in the document]
Comment (miwurster) [reply org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/edmm/EdmmResource.java:235]: Fixed. No just the operation name matters.

**Review by lharzenetter — APPROVED** [review]

**Comment by miwurster** [comment]
CQ has been created at <https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23001>
